src/csharp/Grpc.Core/Internal/UserAgentStringProvider.cs

```diff
@@ -0,0 +1,114 @@
+        }
+
+        public static UserAgentStringProvider DefaultInstance => defaultInstance;
+
+        public string GrpcCsharpUserAgentString => userAgentString;
+
+        public UserAgentStringProvider(string grpcCsharpVersion, string frameworkDescription, string clrVersion, string tfm, CommonPlatformDetection.CpuArchitecture arch)
+        {
+            var detailComponents = new List<string>();
+
+            string sanitizedFrameworkDescription = SanitizeFrameworkDescription(frameworkDescription);
+            if (sanitizedFrameworkDescription != null)
+            {
+                detailComponents.Add(sanitizedFrameworkDescription);
+            }
+
+            if (clrVersion != null)
+            {
+                detailComponents.Add($"CLR {clrVersion}");
+            }
+
+            if (tfm != null)
+            {
+                detailComponents.Add(tfm);
+            }
+
+            string architectureString = TryGetArchitectureString(arch);
+            if (architectureString != null)
+            {
+                detailComponents.Add(architectureString);
+            }
+
+            // TODO(jtattermusch): consider adding details about running under unity / xamarin etc.
+            var details = string.Join("; ", detailComponents);
+            userAgentString = $"grpc-csharp/{grpcCsharpVersion} ({details})";
+        }
+
+        static string TryGetArchitectureString(CommonPlatformDetection.CpuArchitecture arch)
+        {
+            switch (arch)
+            {
+                case CommonPlatformDetection.CpuArchitecture.X86:
+                  return "x86";
+                case CommonPlatformDetection.CpuArchitecture.X64:
+                  return "x64";
+                case CommonPlatformDetection.CpuArchitecture.Arm64:
+                  return "arm64";
+                default:
+                  return null;
+            }
+        }
+
+        static string SanitizeFrameworkDescription(string frameworkDescription)
+        {
+            if (frameworkDescription == null)
+            {
+                return null;
+            }
+
+            // Some platforms return more details in the FrameworkDescription string than we want.
+            // e.g. on mono, we will get something like "Mono 6.12.0.93 (2020-02/620cf538206 Tue Aug 25 14:04:52 EDT 2020)"
+            // For user agent string, we only want basic info on framework name and its version.
+            var parts = new List<string>(frameworkDescription.Split(' '));
+            
+            int i = 0;
+            for  (; i < parts.Count; i++)
+            {
+                var part = parts[i];
+                if (!Regex.IsMatch(part, @"^[-.,+@A-Za-z0-9]*$"))
+                {
+                    // stop once we find first part that's not framework name or version
+                    break;
+                }
+            }
+
+            var result = string.Join(" ", parts.GetRange(0, i));
```

Comment:
nit: move this `string.Join(" ", parts.GetRange(0, i));` assignment to line 105, just before breaking out of the loop? Then, we can keep `i` as a loop-local